lib/Rx/Operator/DelayOperator.php

```diff
@@ -2,81 +2,92 @@
-                    $observer->onNext($x);
-                    $disposable->remove($schedDisp);
-                }, $this->delay);
-
-                $disposable->add($schedDisp);
-            },
-            function ($err) use ($scheduler, $observer, &$lastScheduledTime, $disposable, &$sourceDisposable) {
-                $disposable->remove($sourceDisposable);
-                $sourceDisposable->dispose();
-                $observer->onError($err);
-            },
-            function () use ($scheduler, $observer, $disposable, &$sourceDisposable) {
-                $disposable->remove($sourceDisposable);
-                $sourceDisposable->dispose();
-                $schedDisp = $scheduler->schedule(function () use ($observer, &$schedDisp, $disposable) {
-                    $observer->onCompleted();
-                    $disposable->remove($schedDisp);
-                }, $this->delay);
+        /** @var AnonymousObservable $observable */
+        return $observable
+            ->materialize()
+            ->timestamp()
+            ->map(function (Timestamped $x) {
+                return new Timestamped($x->getTimestampMillis() + $this->delayTime, $x->getValue());
+            })
+            ->subscribe(new CallbackObserver(
+                function (Timestamped $x) use ($scheduler, $observer) {
+                    if ($x->getValue() instanceof Notification\OnErrorNotification) {
+                        $x->getValue()->accept($observer);
+                    }
+                    $this->queue->enqueue($x);
+                    if ($this->schedulerDisposable === null) {
+                        $doScheduledStuff = function () use ($observer, $scheduler, &$doScheduledStuff) {
+                            while ((!$this->queue->isEmpty()) && $scheduler->now() >= $this->queue->bottom()->getTimestampMillis()) {
+                                /** @var Timestamped $item */
+                                $item = $this->queue->dequeue();
+                                /** @var Notification $materializedValue */
+                                $materializedValue = $item->getValue();
+                                $materializedValue->accept($observer);
+                            }
 
-                $disposable->add($schedDisp);
-            }),
-            $scheduler
-        );
+                            if ($this->queue->isEmpty()) {
+                                $this->schedulerDisposable = null;
+                                return;
+                            }
+                            $this->schedulerDisposable = $scheduler->schedule($doScheduledStuff, $this->queue->bottom()->getTimestampMillis() - $scheduler->now());
+                        };
 
-        $disposable->add($sourceDisposable);
+                        $doScheduledStuff();
+                    }
+                },
+                [$observer, 'onError'],
+                function () {
```

Comment:
You can probably just get rid or the onCompleted argument, since it's replaced with a "noop" in `CallbackObserver`